src/cljs/ataru/hakija/application_hakukohde_component.cljs

```diff
@@ -4,37 +4,78 @@
     [ataru.application-common.application-field-common :refer [scroll-to-anchor]]
     [ataru.cljs-util :refer [get-translation]]))
 
-(defn index-of [s val from-index]
-  (clojure.string/index-of (clojure.string/lower-case s)
-                           (clojure.string/lower-case val)
-                           from-index))
-
 (defn- should-search? [search-term]
   (> (count search-term) 1))
 
-(defn match-text [text search-term]
-  (if-not (should-search? search-term)
+(defn text-matches
+  [text search-terms]
+  (let [text         (clojure.string/lower-case text)
+        search-terms (->> search-terms
+                          (filter should-search?)
+                          (map clojure.string/lower-case))]
+    (apply concat
+           (for [search-term search-terms]
+             (loop [res           []
+                    current-index 0]
+               (let [match-index (clojure.string/index-of text search-term current-index)
+                     match-end   (when match-index (+ match-index (count search-term)))]
+                 (if (nil? (and match-index match-end))
+                   res
+                   (recur
+                     (conj res [match-index match-end])
+                     match-end))))))))
+
+(defn- reduce-matches-pass
+  [text-matches]
+  (reduce
+    (fn [acc [match-start match-end]]
+      (let [[previous-start previous-end] (last acc)]
+        (if (and previous-start (<= match-start previous-end))
+          (conj (butlast acc) [previous-start match-end])
+          (conj acc [match-start match-end]))))
+    []
+    (sort text-matches)))
+
+(defn reduce-matches
```

Comment:
Mikäli mun aivot ei petä mua, niin `reduce-matches-pass`:n pitäis hoitaa asia aina yhdellä kertaa, jolloin tätä ei pitäisi tarvita.

Reply:
Tossa on sellanen homma, että "päällekkäiset" matchit vaatii useamman iteraatiot yhdistämiseksi. Eli jos on vaikka matches `([3 5] [4 6] [0 2] [8 10] [3 6])` niin yksi iteraatio palauttaa `([8 10] [4 6] [3 6] [0 2])` ja toinen `([8 10] [3 6] [0 2])`. Vastaavasti noi ylemmät tasot vaatii, ettei ole päällekkäisiä indeksejä, ja onhan se muutenkin nätimpää sievistää nuo loppuun asti. Toteutus voisi varmastikin olla nätimpikin :)